packages/metro/src/ModuleGraph/worker/collectDependencies.js

```diff
@@ -138,6 +140,42 @@ function collectDependencies(
   };
 
   const visitor = {
+    TryStatement(path: Path, state: State) {
+      path.get('block').traverse({
+        CallExpression(p) {
+          const getOptionalDependency = (): string | null => {
```

Comment:
Can you confirm this only evaluates one level deep in try-blocks and it will not work within nested-if statements or function bodies? I think this makes a lot of sense as it reduces complexity somewhat.

Reply:
"one-level deep"? in terms of runtime call-hierarchy or the code-structure? 

- code-structure: every statement within the **lexical** try-block will be parsed (by babel parser), so it is not shallow. 
- call-hierarchy: babel parser is a lexical analyzer so it does not evaluate the statements nor following function calls, so you can say it is shallow in terms of call-stack?

example:

```
function outside() {
  import('not-optional");
}
try {
  import('this-is-optional');
  if(true){
    if(false) {
        import('this-is-also-optional');
    }
  }
  outside();
} catch (e) {}
```

The example above is merely for illustration... usually, an optional dependency is enclosed in a much tighter try-block since it needs very specific "catch" logic to handle the missing module situation ...

Reply:
Can we instead change it to check only for `require` calls directly within one-level of a try-block?

Example:
```
// Valid
let a;
try {
  a = require('module');
} catch {}

// Invalid

try {
  if ('maybeInclude') {
    require('module');
  }
} catch {}
```

You can achieve this by directly iterating over the items in `block`  instead of using `traverse`.

I think we can reduce complexity here – I don't see any case where a try block for an optional dependency should be more complex than assigning a variable.

Reply:
> Can we instead change it to check only for require calls directly within one-level of a try-block?... I think we can reduce complexity here

we could certainly constraint to one-level, but what does 'complexity' mean?

```
try {
  // hundreds of line of code here
  ...
  a = require('module');
  // another hundred of line of code here
  ...
} catch {}
```

The example above is one-level, but is it considered complex? Maybe we should also consider limiting the try-block statement number to be 1 to make it not-complex? But a single statement can also be "complex", so goes the rabbit hole...

On the other hand, if  the goal of trying to "reduce complexity" is to make sure developers won't accidentally put the non-optional dependency in the try-block, then instead of using the level, or any other kinds of complexity measurement to **guess**, I wonder would it be more direct and fool-proof to use an explicit token, such as a comment to safeguard such intention:

```
try {
  // metro: optional-dependency
  a = require('this-is-optional');

  b = require('this-is-NOT-optional');
} catch {}
```

So all the existing dependency in try-block, even when the `allowOptionalDependencies` is turned on, will not be optional, unless developers explicitly added this comment directly above the call... Does that address your concern?

Reply:
I am fine if we limit it to variable assignments one-level deep:

```
try {
  a = require('someModule');
  b = require('someOtherModule');
} catch {}
```

Invalid:
```
try {
  a = require('someModule');
  runSomeCode();
  b = require('someOtherModule');
} catch {}
```

Any other code can be moved out of the try block and handled afterwards. That way we ensure optional dependencies can only be used via a very single pattern – it's an escape hatch with limited API support.

Reply:
ok, I refactor the code to adopt the bottom-up approach and piggyback on the existing import/require expression processing to completely removed the TryStatement handler (top-down). The optional dependency check is now simply to see if the TryStatement is in its "immediate" parent chain after the import/require expression is vetted in the existing CallExpression handlers. 

However, I didn't implement any other sibling level constraint from the "invalid" example above. Sibling level constraint will make otherwise simple and linear parent-child traversal a lot more complicated, because it made the expression condition depends on all the other expressions/statements within the block... it didn't add any feature as far as I can see and certainly didn't further reduce the complexity in the extraction code, our ultimate goal...

Let me know if this makes sense and I will do the final rebase/merge.

Reply:
This sounds great actually. Can you rebase so I can take another look before importing this to FB?